src/services.py

```diff
@@ -514,131 +516,297 @@ def ban_user(self, my_user_id, other_user_id):
+    def __init__(self, config):
+        self.config = config
+
+    @staticmethod
+    def _serialize_offer(offer):
+        return {
+            "id": offer.get("id"),
+            "price": offer.get("price"),
+            "number_of_shares": offer.get("number_of_shares"),
+            "offer_status": offer.get("offer_status"),
+            "created_at": datetime.timestamp(offer.get("created_at")) * 1000,
+            "type": "offer",
+        }
+
+    @staticmethod
+    def _serialize_chat_offer(chat_room_id, offer, is_deal_closed):
+        return {
+            "chat_room_id": chat_room_id,
+            "updated_at": datetime.timestamp(offer.get("created_at")) * 1000,
+            "new_chat": OfferService._serialize_offer(offer=offer),
+            "is_deal_closed": is_deal_closed,
+        }
+
+    @staticmethod
+    def _get_current_offer(session, offer):
+        session.add(offer)
+        session.flush()
+        session.refresh(offer)
+        return offer.asdict()
+
+    @staticmethod
+    def _update_chatroom_datetime(session, chat_room, offer):
+        chat_room.updated_at = offer.get("created_at")
+        session.commit()
+
+    @staticmethod
+    def _verify_user(chat_room, user_id, user_type):
+        if (user_type == "buyer" and chat_room.buyer_id != user_id) or (
+            user_type == "seller" and chat_room.seller_id != user_id
+        ):
+            raise ResourceNotOwnedException("Wrong user")
+
+    @staticmethod
+    def _update_offer_status(session, offer, chat_room, offer_status):
+        chat_room.updated_at = offer.created_at
+        chat_room.is_deal_closed = True
+        offer.offer_status = offer_status
+        session.commit()
+
+    def create_new_offer(
+        self, chat_room_id, author_id, price, number_of_shares, user_type
+    ):
+        with session_scope() as session:
+            chat_room = session.query(ChatRoom).get(chat_room_id)
+            if chat_room == None:
+                raise ResourceNotFoundException("Chat room not found")
+            if chat_room.is_deal_closed == True:
+                raise InvalidRequestException("Deal is closed")
+            OfferService._verify_user(
+                chat_room=chat_room, user_id=author_id, user_type=user_type
+            )
+            offer = Offer(
+                chat_room_id=str(chat_room_id),
+                price=price,
+                number_of_shares=number_of_shares,
+                author_id=str(author_id),
+            )
+            offer = OfferService._get_current_offer(session=session, offer=offer)
+            OfferService._update_chatroom_datetime(
+                session=session, chat_room=chat_room, offer=offer
+            )
+            return OfferService._serialize_chat_offer(
+                chat_room_id=chat_room_id,
+                offer=offer,
+                is_deal_closed=chat_room.is_deal_closed,
+            )
+
+    def accept_offer(self, chat_room_id, offer_id, user_id, user_type):
```

Comment:
This seems to be a lot of duplicates with `reject_offer`, can we abstract them together?